openslides/utils/cache.py

```diff
@@ -1,506 +1,424 @@
+    Cache for the CollectionElements.
 
-    def save_data(self, data: UserCacheDataType) -> None:
-        """
-        Saves the full data set (like created with build_data) to the cache.
-        """
-        raise NotImplementedError()
+    Saves the full_data and if enabled the restricted data.
 
-    def build_data(self) -> UserCacheDataType:
-        """
-        Creates all the data, saves it to the cache and returns it.
-        """
-        websocket_user_ids = defaultdict(set)  # type: UserCacheDataType
-        for channel_name in Group('site').channel_layer.group_channels('site'):
-            session = session_for_reply_channel(channel_name)
-            user_id = session.get('user_id', None)
-            websocket_user_ids[user_id or 0].add(channel_name)
-        self.save_data(websocket_user_ids)
-        return websocket_user_ids
+    There is one redis Hash (simular to python dict) for the full_data and one
+    Hash for every user.
 
-    def get_cache_key(self) -> str:
-        """
-        Returns the cache key.
-        """
-        return self.cache_key
+    The key of the Hashes is COLLECTIONSTRING:ID where COLLECTIONSTRING is the
+    collection_string of a collection and id the id of an element.
 
+    All elements have to be in the cache. If one element is missing, the cache
+    is invalid, but this can not be detected. When a plugin with a new
+    collection is added to OpenSlides, then the cache has to be rebuild manualy.
 
-class RedisWebsocketUserCache(BaseWebsocketUserCache):
-    """
-    Implementation of the WebsocketUserCache that uses redis.
+    There is an sorted set in redis with the change id as score. The values are
+    COLLETIONSTRING:ID for the elements that have been changed with that change
+    id. Which this key it is possible, to get all elements as full_data or as
+    restricted_data that are newer then a specific change id.
 
-    This uses one cache key to store all connected user ids in a set and
-    for each user another set to save the channel names.
+    All method of this class are async. You either have to call them with
+    await in an async environment or use asgiref.sync.async_to_sync().
     """
 
-    def add(self, user_id: int, channel_name: str) -> None:
-        """
-        Adds a channel name to an user id.
-        """
-        redis = get_redis_connection()
-        pipe = redis.pipeline()
-        pipe.sadd(self.get_cache_key(), user_id)
-        pipe.sadd(self.get_user_cache_key(user_id), channel_name)
-        pipe.execute()
-
-    def remove(self, user_id: int, channel_name: str) -> None:
-        """
-        Removes one channel name from the cache.
-        """
-        redis = get_redis_connection()
-        redis.srem(self.get_user_cache_key(user_id), channel_name)
-
-    def get_all(self) -> UserCacheDataType:
-        """
-        Returns all data using a dict where the key is a user id and the value
-        is a set of channel_names.
+    def __init__(
```

Comment:
The init options cachable_provider and start_time are not in use, right? Or do you use this in the tests only? Maybe add a short comment.

Reply:
Of cause thy are in use. Or do you mean, that they always have to same value in production?

Can you give me an example? All I can think of is to obvious
```
# cache_provider_class, cachable_provider and start_time are used be this
# class to access other parts of OpenSlides. You can change them for
# testing.
```
```
cache_provider_class, cachable_provider and start_time are used be this 
class to access other parts of OpenSlides.
```